server/manager/src/main/java/org/apache/accumulo/manager/TabletGroupWatcher.java

```diff
@@ -924,6 +967,135 @@ private void mergeMetadataRecords(MergeInfo info) throws AccumuloException {
+
+    // This covers the case of when a deletion range overlaps the first tablet
+    // We need to create a range that excludes the deletion
+    if ((firstKey == null || !firstKey.equals(deleteRange.toDataRange().getStartKey()))
+        && deleteRange.toDataRange().getStartKey() != null
+        && tabletRange.contains(deleteRange.toDataRange().getStartKey())) {
+      Manager.log.trace(
+          "Fencing tablet at start of deletion range with end row {}, Start deletion row {}",
+          prevTabletEndRow, deleteRange.toDataRange().getStartKey().getRow());
+      ranges.add(new Range(prevTabletEndRow, false,
+          deleteRange.toDataRange().getStartKey().getRow(), false));
+    }
+
+    // This covers the case of when a deletion range overlaps the last tablet
+    // We need to create a range that excludes the deletion.
+    if ((tabletEndRow == null
+        || !tabletEndRow.equals(deleteRange.toDataRange().getEndKey().getRow()))
+        && deleteRange.toDataRange().getEndKey() != null
+        && tabletRange.contains(deleteRange.toDataRange().getEndKey())) {
+      Manager.log.trace(
+          "Fencing tablet at end of deletion range with end row {}, End deletion row {}",
+          tabletEndRow, deleteRange.toDataRange().getEndKey().getRow());
+      ranges.add(new Range(deleteRange.toDataRange().getEndKey().getRow(), tabletEndRow));
+    }
+
+    return ranges;
+  }
+
+  private void updateMetadataRecordsForDelete(MergeInfo info) throws AccumuloException {
+    final KeyExtent range = info.getExtent();
+
+    String targetSystemTable = MetadataTable.NAME;
+    if (range.isMeta()) {
+      targetSystemTable = RootTable.NAME;
+    }
+
+    final AccumuloClient client = manager.getContext();
+
+    try (BatchWriter bw = client.createBatchWriter(targetSystemTable)) {
+      final Text startRow = range.prevEndRow();
+      final Text lastRow = range.endRow() != null
+          ? new Key(range.endRow()).followingKey(PartialKey.ROW).getRow() : null;
+
+      // Find the tablets that overlap the start and end of the deletion range
+      final List<TabletMetadata> tabletMetadatas = new ArrayList<>();
+      loadTabletMetadata(range.tableId(), startRow, ColumnType.PREV_ROW, ColumnType.FILES)
+          .ifPresent(tabletMetadatas::add);
+      loadTabletMetadata(range.tableId(), lastRow, ColumnType.PREV_ROW, ColumnType.FILES)
+          .ifPresent(tabletMetadatas::add);
+
+      for (TabletMetadata tabletMetadata : tabletMetadatas) {
+        final KeyExtent keyExtent = tabletMetadata.getExtent();
+
+        // Check if this tablet needs to have its files fenced for the deletion
+        if (needsFencingForDeletion(info, keyExtent)) {
+          Manager.log.debug("Found overlapping keyExtent {} for delete, fencing files.", keyExtent);
+
+          // Create the ranges for fencing the files, this takes the place of
+          // chop compactions and splits
+          final List<Range> ranges = createRangesForDeletion(tabletMetadata, range);
```

Comment:
Can this return an empty list and if so what should happen?

Reply:
Hmm, I don't think it would ever be empty because we are only iterating over the 1 or 2 tablets that overlap the deletion range by looking up the tablets previously. I would need to think about it but at least in the existing tests it's never empty and should always be at least 1 range or 2 (if the deletion is contained in 1 tablet). We may want to add a check for this to throw an error as I don't think it would be valid if it was empty.

Reply:
One case I was unsure about was what happens when deleting null,null, basically all data.   Was not sure if this code gets involved in that case.

Reply:
That case works and is handled and is covered by DeleteRowsIT. In this case it never hits that block of code because the [line](https://github.com/cshannon/accumulo/blob/64484c420ec5c61e3d7dcfbb47c4b4bcfe88ab16/server/manager/src/main/java/org/apache/accumulo/manager/TabletGroupWatcher.java#L1025) that checks if the tablet needs fencing will be false.

When it runs with null, null the returned tablet that overlaps is `1:a<;` and that fails the needs fencing check so it never processes anything and [logs](https://github.com/cshannon/accumulo/blob/64484c420ec5c61e3d7dcfbb47c4b4bcfe88ab16/server/manager/src/main/java/org/apache/accumulo/manager/TabletGroupWatcher.java#L1084) a statement. No tablets are fenced and later all tablets get deleted. This check ends up being false and you see the following in the log when running a test:

`2023-09-01T17:18:38,785 [manager.Manager] DEBUG: Skipping metadata update on file for keyExtent 1;a< for delete as not overlapping on rows`

Reply:
Thanks for the info it will save me time when I get back to reviewing this, I had not looked at `needsFencingForDeletion` yet.